# REDES NEURONALES

In [2]:
#Importando librerias

import numpy as np
import pandas as pd
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from sklearn.model_selection import train_test_split

Se genero un excel con los datos consolidados que se utilizaron en R para importarlos en este ambiente.

In [11]:
datos_consolidados = pd.read_excel('datos_consolidados.xlsx')

In [12]:
datos_consolidados.head()

núm_corre  año_ocu  mes_ocu  día_ocu  día_sem_ocu  hora_ocu  g_hora  \
0          1     2015        1        1            4        16       3   
1          2     2015        1        1            4        16       3   
2          3     2015        1        1            4        22       4   
3          4     2015        1        1            4         2       1   
4          5     2015        1        1            4         9       2   

   g_hora_5  depto_ocu  mupio_ocu  ...  edad_quinquenales  fall_les  \
0         2          1        101  ...                  4         2   
1         2          1        101  ...                  4         2   
2         3          1        101  ...                  7         2   
3         1          1        101  ...                  4         2   
4         1          1        101  ...                  5         1   

   int_o_noint  tipo_veh  marca_veh  color_veh  modelo_veh  g_modelo_veh  \
0            1         4         21          5        2011             5   
1            1         4         21          5        2011             5   
2            1         4         21          5        9999             6   
3            1         4         21          1        2000             4   
4            2         4         40          5        9999             6   

   tipo_eve  horario  
0         1        1  
1         1        1  
2         1        2  
3         1        2  
4         2        1  

[5 rows x 26 columns]

In [13]:
datos_consolidados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91367 entries, 0 to 91366
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   núm_corre          91367 non-null  int64
 1   año_ocu            91367 non-null  int64
 2   mes_ocu            91367 non-null  int64
 3   día_ocu            91367 non-null  int64
 4   día_sem_ocu        91367 non-null  int64
 5   hora_ocu           91367 non-null  int64
 6   g_hora             91367 non-null  int64
 7   g_hora_5           91367 non-null  int64
 8   depto_ocu          91367 non-null  int64
 9   mupio_ocu          91367 non-null  int64
 10  zona_ocu           91367 non-null  int64
 11  sexo_per           91367 non-null  int64
 12  edad_per           91367 non-null  int64
 13  mayor_menor        91367 non-null  int64
 14  g_edad_80ymás      91367 non-null  int64
 15  g_edad_60ymás      91367 non-null  int64
 16  edad_quinquenales  91367 non-null  int64
 17  fall_les    

In [14]:
datos_consolidados.describe()

núm_corre       año_ocu       mes_ocu       día_ocu   día_sem_ocu  \
count  91367.000000  91367.000000  91367.000000  91367.000000  91367.000000   
mean    5137.297700   2019.032167      6.527072     15.722274      4.356245   
std     3029.321818      2.649056      3.537576      8.862713      2.064672   
min        1.000000   2015.000000      1.000000      1.000000      1.000000   
25%     2538.500000   2017.000000      3.000000      8.000000      3.000000   
50%     5076.000000   2019.000000      7.000000     16.000000      5.000000   
75%     7614.000000   2021.000000     10.000000     23.000000      6.000000   
max    11668.000000   2023.000000     12.000000     31.000000      7.000000   

           hora_ocu        g_hora      g_hora_5     depto_ocu     mupio_ocu  \
count  91367.000000  91367.000000  91367.000000  91367.000000  91367.000000   
mean      13.554215      2.847264      1.993838      7.744131    778.921416   
std        6.849150      1.063200      0.844961      6.961212    695.779349   
min        0.000000      1.000000      1.000000      1.000000    101.000000   
25%        8.000000      2.000000      1.000000      1.000000    109.000000   
50%       15.000000      3.000000      2.000000      5.000000    506.000000   
75%       19.000000      4.000000      3.000000     14.000000   1413.000000   
max       99.000000      5.000000      4.000000     22.000000   2217.000000   

       ...  edad_quinquenales      fall_les   int_o_noint      tipo_veh  \
count  ...       91367.000000  91367.000000  91367.000000  91367.000000   
mean   ...           7.693949      1.819278      1.263235      9.528243   
std    ...           4.332270      0.388047      0.614172     21.587794   
min    ...           1.000000      1.000000      1.000000      1.000000   
25%    ...           5.000000      2.000000      1.000000      3.000000   
50%    ...           6.000000      2.000000      1.000000      4.000000   
75%    ...           9.000000      2.000000      1.000000      4.000000   
max    ...          18.000000      9.000000      9.000000     99.000000   

          marca_veh     color_veh    modelo_veh  g_modelo_veh      tipo_eve  \
count  91367.000000  91367.000000  91367.000000  91367.000000  91367.000000   
mean     303.501220     29.492267   7641.748706     48.684208      2.647477   
std      424.957515     42.279423   3644.142801     46.837587      2.540236   
min        1.000000      1.000000   1898.000000      1.000000      1.000000   
25%       27.000000      2.000000   2019.000000      5.000000      1.000000   
50%       67.000000      5.000000   9999.000000      6.000000      1.000000   
75%      999.000000     99.000000   9999.000000     99.000000      5.000000   
max      999.000000     99.000000   9999.000000     99.000000     99.000000   

            horario  
count  91367.000000  
mean       1.435102  
std        0.495773  
min        1.000000  
25%        1.000000  
50%        1.000000  
75%        2.000000  
max        2.000000  

[8 rows x 26 columns]

## Primer modelo: Predecir tipo de evento

Se realizan los mismos filtros y se trabaja con un modelo de prediccion no binario para comparar los resultados obtenidos con otros modelos.

In [79]:
data_arbol2 = datos_consolidados[
    (datos_consolidados["fall_les"].isin([1, 2])) &
    (datos_consolidados["sexo_per"] != 9) &
    (datos_consolidados["tipo_eve"] != 99) &
    (datos_consolidados["tipo_eve"] != 1) &
       (datos_consolidados["tipo_veh"] != 99) &
    (datos_consolidados["edad_per"] != 999)].copy()


In [80]:
X = data_arbol2[['tipo_veh','depto_ocu','fall_les', 'mayor_menor','horario','sexo_per']]
X = X.fillna(0)
y = data_arbol2['tipo_eve'].values - 1

In [81]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [82]:

model = Sequential()
model.add(Dense(8, input_dim=6, activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(8, activation='softmax'))

C:\Users\Brandon Portillo\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:95: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Esta red contiene 6 variables de entrada y 8 salidas debido a que se trata de 8 casos a analizar.

In [83]:

model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [84]:

model.fit(X_train, y_train, epochs=50, batch_size=64, validation_data=(X_test, y_test))

Epoch 1/50
475/475 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.2494 - loss: 2.0575 - val_accuracy: 0.4397 - val_loss: 1.4917
Epoch 2/50
475/475 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.4459 - loss: 1.4595 - val_accuracy: 0.4490 - val_loss: 1.4519
Epoch 3/50
475/475 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.4509 - loss: 1.4337 - val_accuracy: 0.4470 - val_loss: 1.4276
Epoch 4/50
475/475 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.4523 - loss: 1.4069 - val_accuracy: 0.4500 - val_loss: 1.3981
Epoch 5/50
475/475 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.4514 - loss: 1.3844 - val_accuracy: 0.4499 - val_loss: 1.3795
Epoch 6/50
475/475 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.4515 - loss: 1.3701 - val_accuracy: 0.4523 - val_loss: 1.3719
Epoch 7/50
475/475 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.4565 - loss: 1.3606 - val_accuracy: 0.4544 - val_loss: 1.3673
Epoch 8/50
475/475 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.4639 - loss: 1.3541 - val_accuracy: 0.

In [85]:

loss, accuracy = model.evaluate(X_test, y_test)
print(f'Loss: {loss}, Accuracy: {accuracy}')

238/238 ━━━━━━━━━━━━━━━━━━━━ 0s 851us/step - accuracy: 0.4681 - loss: 1.3195
Loss: 1.3195009231567383, Accuracy: 0.46806269884109497


El modelo no acierta el 46.8% de las veces

In [86]:
caso1= np.array([[2,1,1,4,1,1]])
caso2= np.array([[1,2,2,1,6,1]])
caso3= np.array([[2,1,2,3,14,2]])
caso4= np.array([[1,1,1,2,18,2]])
caso5= np.array([[2,2,1,5,22,2]])
casos=[caso1,caso2,caso3,caso4,caso5]

In [89]:
predicciones_arbol2= []
for i in casos:
    predicciones_arbol2.append(model.predict(i))

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 24ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 30ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 23ms/step


In [90]:
pred_array = np.vstack(predicciones_arbol2)   # queda shape (n, 22)

df_prediccion = pd.DataFrame(
    np.round(pred_array, 3),                 # Redondear a 3 decimales
    index = np.arange(1, pred_array.shape[0] + 1),   # Filas desde 1
    columns = np.arange(1, pred_array.shape[1] + 1)  # Columnas desde 1
)

df_prediccion

1      2      3      4      5      6      7    8
1  0.0  0.213  0.057  0.010  0.708  0.007  0.004  0.0
2  0.0  0.183  0.046  0.006  0.760  0.003  0.001  0.0
3  0.0  0.125  0.017  0.003  0.852  0.003  0.000  0.0
4  0.0  0.119  0.015  0.004  0.858  0.004  0.000  0.0
5  0.0  0.107  0.012  0.004  0.872  0.004  0.000  0.0

## Modelo 2: Prediccion de tipo de vehiculo

In [91]:
data_arbol3 = datos_consolidados[
    (datos_consolidados["fall_les"].isin([1, 2])) &
    (datos_consolidados["sexo_per"] != 9) &
    (datos_consolidados["tipo_eve"] != 99) &
    (datos_consolidados["tipo_veh"] != 4) &
       (datos_consolidados["tipo_veh"] != 99) &
    (datos_consolidados["edad_per"] != 999)].copy()

In [92]:
X = data_arbol3[['tipo_eve','sexo_per','depto_ocu','horario','fall_les', 'mayor_menor']]
X = X.fillna(0)
y = data_arbol3['tipo_veh'].values - 1

In [93]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Se tienen 24 casos de salida y 6 de entrada

In [94]:
model = Sequential()
model.add(Dense(16, input_dim=6, activation='relu'))
model.add(Dense(16, activation='relu'))
model.add(Dense(24, activation='softmax'))

C:\Users\Brandon Portillo\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:95: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [95]:

model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [96]:

model.fit(X_train, y_train, epochs=50, batch_size=64, validation_data=(X_test, y_test))

Epoch 1/50
507/507 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.3342 - loss: 2.0766 - val_accuracy: 0.3490 - val_loss: 1.9612
Epoch 2/50
507/507 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.3546 - loss: 1.9496 - val_accuracy: 0.3591 - val_loss: 1.9361
Epoch 3/50
507/507 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.3583 - loss: 1.9321 - val_accuracy: 0.3594 - val_loss: 1.9240
Epoch 4/50
507/507 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.3599 - loss: 1.9209 - val_accuracy: 0.3628 - val_loss: 1.9150
Epoch 5/50
507/507 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.3605 - loss: 1.9135 - val_accuracy: 0.3644 - val_loss: 1.9106
Epoch 6/50
507/507 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.3633 - loss: 1.9083 - val_accuracy: 0.3637 - val_loss: 1.9015
Epoch 7/50
507/507 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.3637 - loss: 1.9024 - val_accuracy: 0.3636 - val_loss: 1.9051
Epoch 8/50
507/507 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.3644 - loss: 1.8981 - val_accuracy: 0.

In [97]:
loss, accuracy = model.evaluate(X_test, y_test)
print(f'Loss: {loss}, Accuracy: {accuracy}')

254/254 ━━━━━━━━━━━━━━━━━━━━ 0s 852us/step - accuracy: 0.3743 - loss: 1.8548
Loss: 1.8547948598861694, Accuracy: 0.37432098388671875


El modelo acerto el 37% de las veces

In [98]:
caso1= np.array([[2,1,1,2,1,1]])
caso2= np.array([[1,2,2,5,6,1]])
caso3= np.array([[2,1,2,3,14,2]])
caso4= np.array([[1,1,1,1,18,2]])
caso5= np.array([[2,2,1,6,22,2]])
casos=[caso1,caso2,caso3,caso4,caso5]

In [100]:
predicciones_arbol3= []
for i in casos:
    predicciones_arbol3.append(model.predict(i))


1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 25ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 24ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 27ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 28ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step


In [101]:
pred_array = np.vstack(predicciones_arbol3)   # queda shape (n, 22)

df_prediccion = pd.DataFrame(
    np.round(pred_array, 3),                 # Redondear a 3 decimales
    index = np.arange(1, pred_array.shape[0] + 1),   # Filas desde 1
    columns = np.arange(1, pred_array.shape[1] + 1)  # Columnas desde 1
)

df_prediccion

1      2      3    4      5     6      7      8      9      10  ...  \
1  0.518  0.105  0.141  0.0  0.073  0.03  0.021  0.001  0.018  0.004  ...   
2  0.595  0.071  0.032  0.0  0.001  0.00  0.001  0.000  0.148  0.000  ...   
3  0.036  0.001  0.000  0.0  0.000  0.00  0.000  0.000  0.005  0.000  ...   
4  0.004  0.000  0.000  0.0  0.000  0.00  0.000  0.000  0.000  0.000  ...   
5  0.019  0.000  0.000  0.0  0.000  0.00  0.000  0.000  0.000  0.000  ...   

    15   16     17     18   19   20   21     22   23   24  
1  0.0  0.0  0.000  0.007  0.0  0.0  0.0  0.001  0.0  0.0  
2  0.0  0.0  0.001  0.000  0.0  0.0  0.0  0.000  0.0  0.0  
3  0.0  0.0  0.000  0.000  0.0  0.0  0.0  0.000  0.0  0.0  
4  0.0  0.0  0.000  0.000  0.0  0.0  0.0  0.000  0.0  0.0  
5  0.0  0.0  0.000  0.000  0.0  0.0  0.0  0.000  0.0  0.0  

[5 rows x 24 columns]

## Modelo 3: Predecir horario (Dia/Noche)

In [102]:
data_arbol4 = datos_consolidados[
    (datos_consolidados["fall_les"].isin([1, 2])) &
    (datos_consolidados["sexo_per"] != 9) &
    (datos_consolidados["tipo_eve"] != 99) &
       (datos_consolidados["tipo_veh"] != 99) &
    (datos_consolidados["edad_per"] != 999)].copy()

In [103]:
X = data_arbol4[['tipo_eve','sexo_per','tipo_veh','depto_ocu','fall_les', 'mayor_menor']]
X = X.fillna(0)
y = data_arbol4['horario'].values - 1

In [104]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

En este analisis se tienen 2 casos de salida y 6 de entrada

In [105]:
model = Sequential()
model.add(Dense(8, input_dim=6, activation='relu'))
model.add(Dense(4, activation='relu'))
model.add(Dense(2, activation='softmax'))

C:\Users\Brandon Portillo\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:95: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [106]:

model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [107]:

model.fit(X_train, y_train, epochs=50, batch_size=64, validation_data=(X_test, y_test))

Epoch 1/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.5628 - loss: 0.6920 - val_accuracy: 0.5769 - val_loss: 0.6738
Epoch 2/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.5826 - loss: 0.6715 - val_accuracy: 0.5832 - val_loss: 0.6725
Epoch 3/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.5877 - loss: 0.6704 - val_accuracy: 0.5878 - val_loss: 0.6713
Epoch 4/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.5896 - loss: 0.6691 - val_accuracy: 0.5893 - val_loss: 0.6701
Epoch 5/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.5909 - loss: 0.6681 - val_accuracy: 0.5894 - val_loss: 0.6694
Epoch 6/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.5922 - loss: 0.6672 - val_accuracy: 0.5936 - val_loss: 0.6684
Epoch 7/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.5939 - loss: 0.6666 - val_accuracy: 0.5923 - val_loss: 0.6678
Epoch 8/50
997/997 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.5953 - loss: 0.6664 - val_accuracy: 0.

In [108]:
loss, accuracy = model.evaluate(X_test, y_test)
print(f'Loss: {loss}, Accuracy: {accuracy}')

499/499 ━━━━━━━━━━━━━━━━━━━━ 0s 794us/step - accuracy: 0.5983 - loss: 0.6654
Loss: 0.6653771996498108, Accuracy: 0.5982686281204224


El modelo acerto el 59% de las veces

In [109]:
caso1= np.array([[2,1,2,4,1,1]])
caso2= np.array([[1,2,5,1,6,1]])
caso3= np.array([[2,2,3,3,14,2]])
caso4= np.array([[1,1,1,2,18,2]])
caso5= np.array([[2,1,6,5,22,2]])
casos=[caso1,caso2,caso3,caso4,caso5]

In [111]:
predicciones_arbol4= []
for i in casos:
    predicciones_arbol4.append(model.predict(i))

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 35ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 29ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 27ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 28ms/step


In [112]:
pred_array = np.vstack(predicciones_arbol4)   # queda shape (n, 22)

df_prediccion = pd.DataFrame(
    np.round(pred_array, 3),                 # Redondear a 3 decimales
    index = np.arange(1, pred_array.shape[0] + 1),   # Filas desde 1
    columns = np.arange(1, pred_array.shape[1] + 1)  # Columnas desde 1
)

df_prediccion

1      2
1  0.430  0.570
2  0.841  0.159
3  0.980  0.020
4  0.989  0.011
5  0.998  0.002

## Modelo 4: Predecir depto

In [113]:
data_arbol5 = datos_consolidados[
    (datos_consolidados["fall_les"].isin([1, 2])) &
    (datos_consolidados["sexo_per"] != 9) &
    (datos_consolidados["tipo_eve"] != 99) &
       (datos_consolidados["tipo_veh"] != 99) &
       (datos_consolidados["depto_ocu"] != 1) &
    (datos_consolidados["edad_per"] != 999)].copy()

In [114]:
X = data_arbol5[['tipo_eve','sexo_per','tipo_veh','fall_les', 'mayor_menor']]
X = X.fillna(0)
y = data_arbol5['depto_ocu'].values - 1

In [115]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

En este modelo se tienen 22 casos de salida y 5 variables de entrada

In [116]:
model = Sequential()
model.add(Dense(16, input_dim=5, activation='relu'))
model.add(Dense(16, activation='relu'))
model.add(Dense(22, activation='softmax'))

C:\Users\Brandon Portillo\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:95: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [117]:
model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [118]:
model.fit(X_train, y_train, epochs=50, batch_size=64, validation_data=(X_test, y_test))

Epoch 1/50
657/657 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.1175 - loss: 2.9883 - val_accuracy: 0.1305 - val_loss: 2.9496
Epoch 2/50
657/657 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.1310 - loss: 2.9397 - val_accuracy: 0.1305 - val_loss: 2.9404
Epoch 3/50
657/657 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.1321 - loss: 2.9331 - val_accuracy: 0.1321 - val_loss: 2.9353
Epoch 4/50
657/657 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.1312 - loss: 2.9292 - val_accuracy: 0.1299 - val_loss: 2.9331
Epoch 5/50
657/657 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.1306 - loss: 2.9270 - val_accuracy: 0.1306 - val_loss: 2.9306
Epoch 6/50
657/657 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.1317 - loss: 2.9248 - val_accuracy: 0.1295 - val_loss: 2.9290
Epoch 7/50
657/657 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.1317 - loss: 2.9231 - val_accuracy: 0.1290 - val_loss: 2.9268
Epoch 8/50
657/657 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.1314 - loss: 2.9212 - val_accuracy: 0.

In [119]:
loss, accuracy = model.evaluate(X_test, y_test)
print(f'Loss: {loss}, Accuracy: {accuracy}')

329/329 ━━━━━━━━━━━━━━━━━━━━ 0s 831us/step - accuracy: 0.1427 - loss: 2.8971
Loss: 2.8970627784729004, Accuracy: 0.14269407093524933


El modelo acerto el 14% de las veces"

In [120]:
caso1= np.array([[2,1,2,4,1]])
caso2= np.array([[1,2,5,1,1]])
caso3= np.array([[2,2,3,3,2]])
caso4= np.array([[1,1,1,2,2]])
caso5= np.array([[2,1,6,5,2]])
casos=[caso1,caso2,caso3,caso4,caso5]

In [121]:
predicciones_arbol5= []
for i in casos:
    predicciones_arbol5.append(model.predict(i))


1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 59ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 28ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 28ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 24ms/step


In [122]:
pred_array = np.vstack(predicciones_arbol5)   # queda shape (n, 22)

df_prediccion = pd.DataFrame(
    np.round(pred_array, 3),                 # Redondear a 3 decimales
    index = np.arange(1, pred_array.shape[0] + 1),   # Filas desde 1
    columns = np.arange(1, pred_array.shape[1] + 1)  # Columnas desde 1
)

df_prediccion


1      2      3      4      5      6      7      8      9      10  ...  \
1  0.0  0.044  0.051  0.089  0.059  0.081  0.047  0.018  0.042  0.045  ...   
2  0.0  0.037  0.059  0.045  0.212  0.045  0.013  0.007  0.046  0.055  ...   
3  0.0  0.038  0.038  0.072  0.061  0.075  0.039  0.015  0.042  0.036  ...   
4  0.0  0.051  0.057  0.062  0.118  0.072  0.040  0.026  0.088  0.032  ...   
5  0.0  0.028  0.083  0.128  0.044  0.056  0.032  0.005  0.024  0.096  ...   

      13     14     15     16     17     18     19     20     21     22  
1  0.044  0.034  0.060  0.098  0.029  0.014  0.034  0.033  0.039  0.033  
2  0.020  0.023  0.022  0.056  0.064  0.078  0.041  0.033  0.020  0.052  
3  0.069  0.065  0.066  0.110  0.040  0.021  0.030  0.040  0.034  0.033  
4  0.026  0.015  0.022  0.050  0.051  0.043  0.049  0.043  0.022  0.064  
5  0.028  0.027  0.081  0.124  0.013  0.007  0.017  0.018  0.048  0.017  

[5 rows x 22 columns]